src/core/ext/transport/chttp2/transport/hpack_table.h

```diff
@@ -46,41 +47,67 @@
 #endif
 
 /* hpack decoder table */
-typedef struct {
+struct grpc_chttp2_hptbl {
+  static uint32_t entries_for_bytes(uint32_t bytes) {
+    return (bytes + GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD - 1) /
+           GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD;
+  }
+  static constexpr uint32_t kInitialCapacity =
+      (GRPC_CHTTP2_INITIAL_HPACK_TABLE_SIZE + GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD -
+       1) /
+      GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD;
+
+  grpc_chttp2_hptbl() {
+    constexpr uint32_t AllocSize = sizeof(ents[0]) * kInitialCapacity;
+    grpc_mdelem* ents_new = static_cast<grpc_mdelem*>(gpr_malloc(AllocSize));
+    memset(ents_new, 0, AllocSize);
+    ents.reset(ents_new);
+  }
+
   /* the first used entry in ents */
-  uint32_t first_ent;
+  uint32_t first_ent = 0;
   /* how many entries are in the table */
-  uint32_t num_ents;
+  uint32_t num_ents = 0;
   /* the amount of memory used by the table, according to the hpack algorithm */
-  uint32_t mem_used;
+  uint32_t mem_used = 0;
   /* the max memory allowed to be used by the table, according to the hpack
      algorithm */
-  uint32_t max_bytes;
+  uint32_t max_bytes = GRPC_CHTTP2_INITIAL_HPACK_TABLE_SIZE;
   /* the currently agreed size of the table, according to the hpack algorithm */
-  uint32_t current_table_bytes;
+  uint32_t current_table_bytes = GRPC_CHTTP2_INITIAL_HPACK_TABLE_SIZE;
   /* Maximum number of entries we could possibly fit in the table, given defined
      overheads */
-  uint32_t max_entries;
+  uint32_t max_entries = kInitialCapacity;
   /* Number of entries allocated in ents */
-  uint32_t cap_entries;
+  uint32_t cap_entries = kInitialCapacity;
   /* a circular buffer of headers - this is stored in the opposite order to
      what hpack specifies, in order to simplify table management a little...
      meaning lookups need to SUBTRACT from the end position */
-  grpc_mdelem* ents;
-  grpc_mdelem static_ents[GRPC_CHTTP2_LAST_STATIC_ENTRY];
-} grpc_chttp2_hptbl;
+  grpc_core::UniquePtr<grpc_mdelem[], grpc_core::DefaultDelete<grpc_mdelem>>
```

Comment:
similar to strings that use `UniquePtr<char>`, I would just use `grpc_core::UniquePtr<grpc_mdelem>` here too. It basically is an array.

Reply:
I don't think that will work if we need to call the `grpc_mdelem` dtor for every element in the array.

Reply:
That's a great point, but it's trivially destructible. 

Sorry I'm missing something but, AFAICT, `grpc_core::DefaultDelete<grpc_mdelem>` would call `([]arr)->~grpc_mdelem`, which I think would map to `arr[0].~grpc_medelem`, right?

Reply:
Sure, but we need it to also call `~grpc_mdelem()` on every element of the array.

Arjun, can you please put together a test that proves that when `UniquePtr<>` is used on an array, that it will call the dtor of every element of the array?  Thanks!

Reply:
Ah ok, sure that makes sense.  Then, I'd like to make sure that it's actually called. I'll wait for @arjunroy to confirm that.

Reply:
We do not need to worry about this.

1) The only time the memory gets de-allocated is if we're re-hashing the table on grow(), or freeing the table at the end with destroy().
2) If we're growing, we explicitly do not want to call the destructor.
3) If we're destroying, we already call Unref() on the metadata before we free the memory.

Reply:
The only value of using a smart pointer here is so that it can take care of the memory management for us.  If we're doing this all manually, then there's no point in using `UniquePtr<>` here.  I think we should go back to the original code where we did the allocation manually.

Reply:
Went back to regular pointers on request.